pkg/controllers/management/nodetemplate/nodetemplate.go

```diff
@@ -0,0 +1,236 @@
+	}
+	ntDynamicClient := dynamicClient.Resource(s)
+
+	nt := nodeTemplateController{
+		ntClient:        mgmt.Management.NodeTemplates(""),
+		ntLister:        mgmt.Management.NodeTemplates("").Controller().Lister(),
+		npLister:        mgmt.Management.NodePools("").Controller().Lister(),
+		npClient:        mgmt.Management.NodePools(""),
+		nsLister:        mgmt.Core.Namespaces("").Controller().Lister(),
+		nsClient:        mgmt.Core.Namespaces(""),
+		nodesLister:     mgmt.Management.Nodes("").Controller().Lister(),
+		nodeClient:      mgmt.Management.Nodes(""),
+		ntDynamicClient: ntDynamicClient,
+		mgmtCtx:         mgmt,
+	}
+
+	mgmt.Management.NodeTemplates("").Controller().AddHandler(ctx, "nt-grb-handler", nt.sync)
+}
+
+func (nt *nodeTemplateController) sync(key string, nodeTemplate *v3.NodeTemplate) (runtime.Object, error) {
+	if nodeTemplate == nil || nodeTemplate.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	annotations := nodeTemplate.Annotations
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	var templateName = nodeTemplate.Name
+	var migratedTemplate bool
+
+	// if owner bindings annotation is present, the node template is in the proper namespace and has had
+	// its creator rolebindings created
+	if ownerBindingsCreated := annotations[ownerBindingsAnno]; ownerBindingsCreated == "true" {
+		return nil, nil
+	}
+
+	creatorID, ok := annotations[rbac.CreatorIDAnn]
+	if !ok {
+		return nodeTemplate, fmt.Errorf("nodeTemplate [%v] has no creatorId annotation", nodeTemplate.Name)
+	}
+
+	if nodeTemplate.Namespace != namespace.NodeTemplateGlobalNamespace {
+		// node template must be migrated to global namespace
+		if err := nt.migrateNodeTemplate(nt.ntDynamicClient, nodeTemplate); err != nil {
+			return nil, err
+		}
+		templateName = fmt.Sprintf("nt-%s-%s", nodeTemplate.Namespace, nodeTemplate.Name)
+		migratedTemplate = true
+	}
+
+	// Create Role and RBs if they do not exist
+	if err := rbac.CreateRoleAndRoleBinding(rbac.NodeTemplateResource, nodeTemplate.Name, namespace.NodeTemplateGlobalNamespace,
+		rbac.RancherManagementAPIVersion, creatorID, []string{rbac.RancherManagementAPIVersion},
+		nodeTemplate.UID,
+		[]v3.Member{}, nt.mgmtCtx); err != nil {
+		return nil, err
+	}
+
+	dynamicNodeTemplate, err := nt.ntDynamicClient.Namespace(namespace.NodeTemplateGlobalNamespace).Get(templateName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	annotations = dynamicNodeTemplate.GetAnnotations()
+	// owner bindings annotation is meant to prevent bindings from being created again if they have been removed from creator
+	annotations[ownerBindingsAnno] = "true"
+	dynamicNodeTemplate.SetAnnotations(annotations)
+
+	if _, err = nt.ntDynamicClient.Namespace(namespace.NodeTemplateGlobalNamespace).Update(dynamicNodeTemplate, metav1.UpdateOptions{}); err != nil {
+		return nil, err
+	}
+
+	// if a migrations has been performed, then the original node template object has not been updated
```

Comment:
If we migrate doesn't that mean the original obj that came into the sync is deleted?